File: tests/Aiko.Core.Tests/SnapshotNameTests.cs
using Aiko.Core;

namespace Aiko.Core.Tests;

public class SnapshotNameTests
{
    [Theory]
    [InlineData(@"C:\Users\someone\.claude-personal", "claude-personal")]
    [InlineData(@"C:\Users\someone\.claude", "claude")]
    [InlineData(@"C:\Users\someone\.claude-work\", "claude-work")]
    [InlineData("/home/someone/.claude", "claude")]
    public void The_file_is_named_after_the_config_folder(string directory, string expected)
    {
        Assert.Equal(expected, SnapshotName.For(directory));
    }

    [Theory]
    [InlineData(null)]
    [InlineData("")]
    [InlineData("   ")]
    public void Without_the_variable_the_name_is_the_default_one(string? directory)
    {
        Assert.Equal(SnapshotName.Default, SnapshotName.For(directory));
    }

    [Fact]
    public void Characters_that_do_not_belong_in_a_file_name_are_dropped()
    {
        Assert.Equal("claudework", SnapshotName.Clean("claude:work?"));
    }

    [Fact]
    public void A_folder_named_in_Russian_keeps_its_letters()
    {
        // The user's home folder can be in any language, and the file name follows it.
        Assert.Equal("клод", SnapshotName.Clean(".клод"));
    }

    [Fact]
    public void A_name_left_with_nothing_falls_back()
    {
        Assert.Equal(SnapshotName.Default, SnapshotName.Clean("..."));
        Assert.Equal(SnapshotName.Default, SnapshotName.Clean("???"));
    }
}

File: mac/Tests/AikoKitTests/SnapshotNameTests.swift
import Testing

@testable import AikoKit

struct SnapshotNameTests {
    @Test(arguments: [
        (#"C:\Users\someone\.claude-personal"#, "claude-personal"),
        (#"C:\Users\someone\.claude"#, "claude"),
        (#"C:\Users\someone\.claude-work\"#, "claude-work"),
        ("/home/someone/.claude", "claude"),
    ] as [(String, String)])
    func theFileIsNamedAfterTheConfigFolder(directory: String, expected: String) {
        #expect(SnapshotName.forConfigDirectory(directory) == expected)
    }

    @Test(arguments: [nil, "", "   "] as [String?])
    func withoutTheVariableTheNameIsTheDefaultOne(directory: String?) {
        #expect(SnapshotName.forConfigDirectory(directory) == SnapshotName.default)
    }

    @Test
    func charactersThatDoNotBelongInAFileNameAreDropped() {
        #expect(SnapshotName.clean("claude:work?") == "claudework")
    }

    @Test
    func aFolderNamedInRussianKeepsItsLetters() {
        // The user's home folder can be in any language, and the file name follows it.
        #expect(SnapshotName.clean(".клод") == "клод")
    }

    @Test
    func aNameLeftWithNothingFallsBack() {
        #expect(SnapshotName.clean("...") == SnapshotName.default)
        #expect(SnapshotName.clean("???") == SnapshotName.default)
    }
}

/// The snapshot name is part of the files already on people's disks. The bridge writes a file under
/// this name and the tray looks for the same name, so these cases must never change.
struct SnapshotNamePinTests {
    @Test(arguments: [
        (#"C:\Users\someone\.claude"#, "claude"),
        (#"C:\Users\someone\.claude-work"#, "claude-work"),
        (#"C:\Users\Александр Иванов\.claude-osnovnaya"#, "claude-osnovnaya"),
        (#"D:\Claude Accounts\.claude-work-2"#, "claude-work-2"),
    ] as [(String, String)])
    func aTypicalWindowsFolderKeepsTheNameItHasToday(directory: String, expected: String) {
        #expect(SnapshotName.forConfigDirectory(directory) == expected)
    }

    @Test(arguments: [
        #"C:\a\.claude-work"#,
        "C:/a/.claude-work",
        #"C:\a\.claude-work\"#,
        "C:/a/.claude-work/",
        #"C:\a/.claude-work\\"#,
    ])
    func theSameFolderGetsTheSameNameWhateverSeparatorsItIsWrittenWith(directory: String) {
        #expect(SnapshotName.forConfigDirectory(directory) == "claude-work")
    }

    @Test(arguments: [#"C:\"#, "C:", #"\"#])
    func aDriveOrARootWithNoFolderNameGetsTheDefaultName(directory: String) {
        #expect(SnapshotName.forConfigDirectory(directory) == SnapshotName.default)
    }

    @Test
    func aFolderOnANetworkShareIsNamedAfterItsLastPart() {
        #expect(SnapshotName.forConfigDirectory(#"\\server\share\users\someone\.claude-team"#) == "claude-team")
    }
}
